Macro

kozusznik opened the pull request.

velissarious committed:
add simple main test to start the plugin

MKrumnikl committed:
Changed template ID to 8 (now requires to upload mpitest.txt).
Remotely executing mpirun
/scratch/work/project/open-15-12/apps/Fiji.app-macro-mpitest/ImageJ-linux64
--headless --console -macro mpitest.txt

MKrumnikl committed:
Added NUMBER_OF_NODES but missing UI setter for it

velissarious committed:
add MPI task progress view. 

New view and controller for MPI tasks progress in Job Detail/Dashboard. Currently it does nothing.

velissarious committed:
added checks if the settings are null

Currently the initialization of the form with dummy values (ex localhost and port 22) and setting the form values equal to the previous form values prevent the settings from ever being null. Therefore the checks result to always being true.

velissarious committed:
disable download filter

Filter is disabled in order to be able to download the output of MPI jobs.

velissarious committed:
add table data on MPI progress view

Currently, the view will display mockup data.

velissarious committed:
add progress indicators in MPI progress view

velissarious committed:
cleanup and correct value for the progress indicator

velissarious committed:
add set method for setting the progress for a specific node

Also no progress indicator is displayed in case that the progress is negative.

velissarious committed:
use constant NUMBER_OF_NODES in MPI progress view

MKrumnikl committed:
Added configuration for number of nodes to UI NewJobView

velissarious committed:
add read from progress_0.plog

In the MPI controller the progress_0.plog file is read.
This will be modified to be usefull.
The BenchmarkJobManager now has a get method that returns Job.
The ObservableBenchmarkJob has a get method that returns Job.
The MPITaskProgressViewController has a set method that sets the selected Job.

velissarious committed:
display real progress from progress files on the MPI Progress view

velissarious committed:
fix different tasks in each node will be read correctly

Each node can define its own tasks (including different ones and omitting other tasks), tasks with the same description are considered to be the same overall task.

MKrumnikl committed:
Added configuration for type of workflow to UI NewJobView. User can
choose SPIM or Macro.

MKrumnikl committed:
Merge branch 'macro' of git@github.com:fiji-hpc/hpc-workflow-manager.git into macro

velissarious committed:
Update MPI task progress every second

velissarious committed:
fix read the number of nodes used from the progress file

MKrumnikl committed:
Fixed UI NewJobView to allow allocation of only single node for SPIM
workflow.

velissarious committed:
fix bug that prevented displaying progress of MPI tasks with nodes > 2

velissarious committed:
Merge remote-tracking branch 'origin/macro' into macro

velissarious committed:
add close for MPI Progress View Controller in JobDetailControl

velissarious committed:
fix bug which would create useless elements in map

velissarious committed:
replaced new runnable with lambda function

MKrumnikl committed:
Added haas_template_id to JobInfo (saved in .jobinfo file)

MKrumnikl committed:
Displaying workflow type in main window.

velissarious committed:
progress will only be set if it is an increase

Progress in an MPITask is now only able to increase.
Other changes in controller, update now does work.

velissarious committed:
Progress tab is enabled according to workflow type

The progress tabs are enabled or disabled according to workflow type.

Made some changes suggested by SonarLint to the Job Detail Controller.

velissarious committed:
clean code

Reduced cognitive complexity of parseProgressLogs in MPI task controller according to SonarLint.

Other clean code modifications according to SonarLint.

velissarious committed:
disable progress modal dialog for MPI tasks

MKrumnikl committed:
Added context menu option for editing macro file.

MKrumnikl committed:
Fix added default macro file to constants.

velissarious committed:
enable download checks, disable snakemake output tab for macro type

Download checks are enabled again, this time there is also a check if the workflow type is macro, in which case it will allow download.

Tab "Snakemake output" is disabled if job is of workflow type macro.

velissarious committed:
clean code

Made some of the changes suggested by SonarLint in BenchmarkJobManager.

velissarious committed:
rename MPI to Macro, dynamically remove irrelevant workflow tabs

Rename all references to MPI to Macro. This is done for naming consistency.

Remove irrelevant to the selected Workflow tabs from job details.

Do some cleanup of the code.

velissarious committed:
clean code

Made some changes according to SonarLint suggestions.

velissarious committed:
change constants interface into a class

velissarious committed:
convert some lambdas to method references

velissarious committed:
rename variables to match camelCase

velissarious committed:
clean code

velissarious committed:
fix: download works for Macro WorkFlow type

This is still incomplete as when the download has finished working the isDone is still not set as true, as a result when the download has finished instead of displaying "done" in the specific row in the download column an empty string will be displayed.

Additional cleanup of the code.

velissarious committed:
fix: progress indicator can only increment

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/BaseExceptionHandler.java:
```diff
@@ -47,7 +47,7 @@ public boolean test(final Thread t, final Throwable exc) {
 			if (closeable != null) {
 				if (log.isDebugEnabled()) {
-					log.debug("Dispose window " + closeable);
+					log.debug("Dispose window ".concat(closeable.toString()));
```
What is an advantage of method concat?

velissarious replied:
Fixed.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/HaasOutputObservableValueRegistry.java:
```diff
@@ -22,8 +21,8 @@
 	private final BenchmarkJob job;
-	private final Map<SynchronizableFileType, SimpleObservableValue<String>> observableValues =
-		new HashMap<>();
+	private final EnumMap<SynchronizableFileType, SimpleObservableValue<String>> observableValues =
```
Type Map is better for variable declaration. Usin EnumMap as implementation is OK.

velissarious replied:
Made the change.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/JobStateComparator.java:
```diff
@@ -2,22 +2,20 @@
 public class JobStateComparator implements Comparator<JobState> {
 
-	private static Map<JobState, Integer> priorities = new HashMap<>();
+	private static EnumMap<JobState, Integer> priorities = new EnumMap<>(JobState.class);
```
Already mentioned.

velissarious replied:
Made the change.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/ObservableBenchmarkJob.java:
```diff
@@ -216,5 +220,9 @@ private void setDone(boolean val) {
 		}
 	}
+	
+	public Job getJob() {
```
Consider using Demeter Law - instead of providing the whole job add an method getWorkflowJobType

velissarious replied:
Done.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/SPIMComputationAccessorDecoratorWithTimeout.java:
```diff
@@ -16,7 +16,8 @@
 	private final  P_ResultCacheHolder<List<String>> outputCache;
 	private final P_ResultCacheHolder<Set<String>> changedFilesCache;
-	private final Map<SynchronizableFileType, Integer> allowedTypesIndices = new HashMap<>();
+	private final EnumMap<SynchronizableFileType, Integer> allowedTypesIndices =
```
Already mentioned.

velissarious replied:
Made the change.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/ui/BenchmarkSPIMControl.java:
```diff
@@ -353,29 +361,28 @@ private void updateJobs(boolean showProgress) {
 			return "Failed";
-		}
-		if (!progress.isWorking() && !progress.isDone()) {
+		} else	if (!progress.isWorking() && !progress.isDone()) {
```
Every if returns from method - construction else is useless.

velissarious replied:
Fixed by adding a temporary variable for the status message which is returned at the end.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/ui/JobDetailControl.java:
```diff
@@ -156,29 +166,35 @@ public void init(final Window parameter) {
-						log.info("Job " + job.getValue().getId() +
-							" state has been resolved as Disposed.");
-					}
+					log.info("Job " + job.getValue().getId() +
```
Consider using https://www.slf4j.org/apidocs/org/slf4j/Logger.html#info(java.lang.String,java.lang.Object) method instead of isInfoEnabled, operator "+" and https://www.slf4j.org/apidocs/org/slf4j/Logger.html#info(java.lang.String)

velissarious replied:
Fixed.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/ui/JobStateNameProvider.java:
```diff
@@ -1,20 +1,19 @@
 class JobStateNameProvider {
 
-	private final Map<JobState,String> translateTable = new HashMap<>();
+	private final EnumMap<JobState,String> translateTable = new EnumMap<>(JobState.class);
```
Already mentioned.

velissarious replied:
Made the change.

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/ui/NewJobController.java:
```diff
@@ -30,6 +29,30 @@
 	}
+	
+	public enum WorkflowType {
+		SPIM_WORKFLOW(4), MACRO_WORKFLOW(8);
```
I am not sure whether it is good to create a hard-coded relation between a template id and a workflow type. It is possible to have different template IDs for a same workflow type.

velissarious replied:
The WorkflowType is read from a file where it is saved as a long id.

kozusznik reviewed (commented):
Please consider and address included notes.

velissarious committed:
fix: EnumMap declaration to Map

kozusznik commented on hpc-workflow-manager-client/src/main/java/cz/it4i/fiji/haas_spim_benchmark/core/MPIComputationAccessorAdapter.java:
```diff
@@ -0,0 +1,36 @@
+
+import cz.it4i.fiji.haas_java_client.SynchronizableFileType;
+
+public class MPIComputationAccessorAdapter implements SPIMComputationAccessor {
```
What is different in comparison to SPIMComputationAccessorAdapter?

velissarious replied:
This file has been removed in a later commit.

kozusznik replied:
OK

kozusznik reviewed (commented):
Another notes

velissarious committed:
fix: Applied Demeter Law, other refactoring

velissarious committed:
fix: disable CONFIG_YAML upload for Macro

kozusznik committed:
fix: Report time during downloading.

velissarious committed:
fix live progress update

kozusznik committed:
fix: Prevent throwing NPE during download due the missing notifier

velissarious committed:
add common macro source code that will be appended to user created macro

The common macro source code required to use MPI that will be appended to the user's macro script on upload.

velissarious committed:
feature: append common macro source code to user's script

On upload the user's script "user.ijm" is copied to a new file "mpitest.txt" along with the commond MPI wrapper code (MacroWrapper.ijm file in the project's resources).

velissarious committed:
feature: if the job is Macro check that there is a user script uploaded

If there is no script uploaded then the Start job item will not appear in the context menu. Currently the method responsible for doing this calls the "getFileContents" method which is slow. This will be replaced with a check to job.upload == true and a check that the user has uploaded a DEFAULT_MACRO_FILE file.

velissarious committed:
fix: check if user has uploaded and script existed on upload fast

This commit replaces the slow "checkMacroScriptIsUploaded" with the much faster "checkIfAnythingHasBeenUploaded" and modified "createTheMacroScript" which now returns boolean which is used to check whether or not to upload the data.
